api/ocm/extensions/accessmethods/git/method.go

```diff
@@ -0,0 +1,188 @@
+	TypeV1 = Type + runtime.VersionSeparator + "v1"
+)
+
+func init() {
+	accspeccpi.RegisterAccessType(accspeccpi.NewAccessSpecType[*AccessSpec](Type, accspeccpi.WithDescription(usage)))
+	accspeccpi.RegisterAccessType(accspeccpi.NewAccessSpecType[*AccessSpec](TypeV1, accspeccpi.WithFormatSpec(formatV1), accspeccpi.WithConfigHandler(ConfigHandler())))
+}
+
+// AccessSpec describes the access for a GitHub registry.
+type AccessSpec struct {
+	runtime.ObjectVersionedType `json:",inline"`
+
+	// RepoURL is the repository URL
+	RepoURL string `json:"repoUrl"`
+
+	// Ref defines the hash of the commit
+	Ref string `json:"ref"`
+
+	// PathSpec is a path in the repository to download, can be a file or a regex matching multiple files
+	PathSpec string `json:"pathSpec"`
+}
+
+// AccessSpecOptions defines a set of options which can be applied to the access spec.
+type AccessSpecOptions func(s *AccessSpec)
+
+// New creates a new git registry access spec version v1.
+func New(url, ref string, pathSpec string, opts ...AccessSpecOptions) *AccessSpec {
+	s := &AccessSpec{
+		ObjectVersionedType: runtime.NewVersionedTypedObject(Type),
+		RepoURL:             url,
+		Ref:                 ref,
+		PathSpec:            pathSpec,
+	}
+	for _, o := range opts {
+		o(s)
+	}
+	return s
+}
+
+func (a *AccessSpec) Describe(internal.Context) string {
+	return fmt.Sprintf("git commit %s[%s]", a.RepoURL, a.Ref)
+}
+
+func (*AccessSpec) IsLocal(internal.Context) bool {
+	return false
+}
+
+func (a *AccessSpec) GlobalAccessSpec(accspeccpi.Context) accspeccpi.AccessSpec {
+	return a
+}
+
+func (*AccessSpec) GetType() string {
+	return Type
+}
+
+func (a *AccessSpec) AccessMethod(c internal.ComponentVersionAccess) (internal.AccessMethod, error) {
+	return accspeccpi.AccessMethodForImplementation(newMethod(c, a))
+}
+
+func newMethod(componentVersionAccess internal.ComponentVersionAccess, accessSpec *AccessSpec) (accspeccpi.AccessMethodImpl, error) {
+	u, err := giturls.Parse(accessSpec.RepoURL)
+	if err != nil {
+		return nil, errors.ErrInvalidWrap(err, "repository repoURL", accessSpec.RepoURL)
+	}
+	if err := plumbing.ReferenceName(accessSpec.Ref).Validate(); err != nil {
+		return nil, errors.ErrInvalidWrap(err, "commit hash", accessSpec.Ref)
+	}
+
+	creds, cid, err := getCreds(accessSpec.RepoURL, componentVersionAccess.GetContext().CredentialsContext())
+	if err != nil {
+		return nil, fmt.Errorf("failed to get credentials for repository %s: %w", accessSpec.RepoURL, err)
+	}
+
+	auth, err := techgit.AuthFromCredentials(creds)
+	if err != nil && !errors.Is(err, techgit.ErrNoValidGitCredentials) {
+		return nil, fmt.Errorf("failed to get auth method for repository %s: %w", accessSpec.RepoURL, err)
+	}
+
+	gitDownloader := git.NewDownloader(u.String(), accessSpec.Ref, accessSpec.PathSpec, auth)
+	cachedGitBlobAccessor := accessobj.CachedBlobAccessForWriter(
+		componentVersionAccess.GetContext(),
+		mime.MIME_OCTET,
+		accessio.NewWriteAtWriter(gitDownloader.Download),
```

Comment:
This could be problematic, If the provided byte stream really always reproducible (or stable)???

If possibly not, the signing would break, if the generic blob digester is used.


There is already an archive digester providing a reproducible git hash.  Therefore, we would require a dedicated mimetype. Because this scenario wasn't used so far. This has not been finally implemented. I just implemented the basic hash function, so far.

Reply:
Changed to use non cached blob access